onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc

```diff
@@ -0,0 +1,242 @@
+
+  // Grab pointers to buffers each Tensor represents:
+  const uint8_t* word_embedding_data = word_embedding->template Data<uint8_t>();
+  const uint8_t* position_embedding_data = position_embedding->template Data<uint8_t>();
+  const uint8_t* segment_embedding_data =
+      has_segment_embedding ? segment_embedding->template Data<uint8_t>() : nullptr;
+  const uint8_t* gamma_data = gamma->template Data<uint8_t>();
+  const uint8_t* beta_data = beta->template Data<uint8_t>();
+
+  T* output_data = output->template MutableData<T>();
+
+  // Perform the Op:
+  {
+    std::atomic_bool failed{false};
+
+    int n = batch_size * sequence_length;
+    concurrency::ThreadPool::TryBatchParallelFor(
+        context->GetOperatorThreadPool(), n, [=, &failed](ptrdiff_t index) {
+          int word_col_index = input_ids_data[index];
+          if (word_col_index < 0 || word_col_index >= word_embedding_length) {
+            failed.store(true, std::memory_order_release);
+            return;
+          }
+          int position_col_index = index % sequence_length;
+          if (position_col_index >= position_embedding_length) {
+            failed.store(true, std::memory_order_release);
+            return;
+          }
+          int segment_col_index = 0;
+          if (nullptr != segment_ids_data) {
+            segment_col_index = segment_ids_data[index];
+            if (segment_col_index < 0 || segment_col_index >= segment_embedding_length) {
+              failed.store(true, std::memory_order_release);
+              return;
+            }
+          }
+
+          // Grab inputs for the embeddings for the current batch index:
+          const uint8_t* input_word_embedding = word_embedding_data + word_col_index * hidden_size;
+          const uint8_t* input_position_embedding =
+              position_embedding_data + position_col_index * hidden_size;
+          const uint8_t* input_segment_embedding = nullptr;
+          if (segment_embedding_data != nullptr) {
+            input_segment_embedding = segment_embedding_data + segment_col_index * hidden_size;
+          }
+
+          T* output = output_data + (index * hidden_size);
+
+          T sum = static_cast<T>(0);
+          for (int i = 0; i < hidden_size; ++i) {
+            T subtotal = Dequantize<uint8_t>(input_word_embedding[i],
+                                             word_embedding_scale,
+                                             word_embedding_zero_point) +
+                         Dequantize<uint8_t>(input_position_embedding[i],
+                                             position_embedding_scale,
+                                             position_embedding_zero_point);
+            if (segment_embedding_data != nullptr) {
+              subtotal += Dequantize<uint8_t>(input_segment_embedding[i],
+                                              segment_embedding_scale,
+                                              segment_embedding_zero_point);
```

Comment:
Yes - ideally the internals could be quantized in a future pass. That was my approach when I started this kernel. Since our partner team is using `quantize_dynamic()` instead of QAT I took this approach to get going :))